pkg/sync/rpc/server.go

```diff
@@ -1505,3 +1529,137 @@ func (b *BackupList) Delete(backupID string) error {
 	}
 	return fmt.Errorf("backup not found %v", backupID)
 }
+
+func (s *SyncAgentServer) SnapshotHash(ctx context.Context, req *ptypes.SnapshotHashRequest) (*empty.Empty, error) {
+	if len(req.SnapshotNames) == 0 {
+		return nil, fmt.Errorf("snapshot name(s) is required")
+	}
+
+	status := hash.DoSnapshotHash(req.SnapshotNames, req.Rehash, req.ConcurrentLimit)
+
+	if err := s.SnapshotHashList.Delete(req.SnapshotNames[0]); err != nil {
+		return nil, errors.Wrapf(err, "failed to delete snapshot %v for from list", req.SnapshotNames[0])
+	}
+
+	if err := s.SnapshotHashList.Add(req.SnapshotNames[0], status); err != nil {
+		return nil, errors.Wrapf(err, "failed to add snapshot %v for hashing", req.SnapshotNames[0])
+	}
+
+	return &empty.Empty{}, nil
+}
+
+func (s *SyncAgentServer) SnapshotHashStatus(ctx context.Context, req *ptypes.SnapshotHashStatusRequest) (*ptypes.SnapshotHashStatusResponse, error) {
+	if req.SnapshotName == "" {
+		return nil, fmt.Errorf("snapshot name is required")
+	}
+
+	task, err := s.SnapshotHashList.Get(req.SnapshotName)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ptypes.SnapshotHashStatusResponse{
+		State:    string(task.State),
+		Progress: int32(task.Progress),
+		Checksum: task.Checksum,
+		Error:    task.Error,
+	}, nil
+}
+
+func (s *SnapshotHashList) Add(snapshotName string, status *replica.SnapshotHashStatus) error {
+	if snapshotName == "" {
+		return fmt.Errorf("snapshot name is required")
+	}
+
+	if err := s.Refresh(); err != nil {
+		return err
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	s.tasks = append(s.tasks, &SnapshotHashInfo{
+		snapshotName: snapshotName,
+		status:       status,
+	})
+
+	return nil
+}
+
+func (s *SnapshotHashList) Refresh() error {
+	s.Lock()
+	defer s.Unlock()
+
+	var index, completed int
+
+	for index = len(s.tasks) - 1; index >= 0; index-- {
```

Comment:
Is this workflow actually a pulling mode notification? Typically we will use the pushing mode for this kind of progress update.

Reply:
Yes, it's a pulling mode. Pulling mode is better. Do we have any pulling mode example in longhorn-engine?

Reply:
I think the current progress update related implementation (backup creation for example) is more like the pushing mode rather than the pulling mode. (The progress and state will be updated inside the progress update function instead of relying on the sync agent server to pull it periodically.)

Reply:
Got it. So, the current method is same as the implemtaion of backup progress update.